### DISCUSSION TITLE
Gpt-engineer abstracted

### DISCUSSION THREAD

**Discussion by RareMojo** [AntonOsika/gpt-engineer#255] [Show and tell]
I was adding a lot of features to the main project. The next morning or so when I'd get around to maybe submitting a PR, I'd notice most of my work already implemented or in progress. 

So I decided I'd start working in a different direction. I wanted to make a more general engineer. 

This adds a new layer called agents, which can have an identity. The identity is created at the start of a run. It uses a template of prompts and fulfills placeholders rather than a single string prompt. This allows your details to be submitted to all of the separate identities used for that agent, allowing for, in theory, any sort of project. Here are some videos showcasing their new abilities as well as showing it can still do the original goal of making a piece of runnable software. 

For the recorded tests for the programming attempt, it did two versions of Snake that were executed immediately without problem. Snake version 2 did require me to go out and find a font.tff that was programmed into the game. However, it launched perfectly as shown after. Snake 1 would have been in the one-shot video, but unfortunately I realized I was capturing the window, rather than my entire display. So it is shown as its own separate video instead.

All project files get created at the end of a run. I do not feel comfortable allowing it to execute yet, but it absolutely could. If you want to see the project or talk about what I did, just let me know. 

I also do not have access to a GPT-4 key. I'm sure if I did, I would have more fun making it code things... lol.
I also realize this may not be the direction this project wants to go. I would agree making it dedicated to programming could be far better than having trade-offs of being abstracted in this way. I think it could probably have best of both worlds or two version or something. Regardless, that's why I formatted this as a discussion instead of sending the code.

Hope you liked looking at what I did.
Have a great day!

YouTube Playlist showcasing the changes:

https://youtube.com/playlist?list=PLSZ76r5stknIx8R6Zt1hRh5nrX0TQLe60

**Comment by noxiouscardiumdimidium** [comment]
you've just described the core functionality of gpt-engineer, that's exactly how it  ALREADY does what it does. and user controlled execution/exit is already implemented as well.... you should probably run `pip install gpt-engineer ; git pull`   you seem to be a week or more out of date, and the past 2 days gits.... GOD-TIER!! ive pulled 3 times TODAY, and it somehow got better between revisions xD

**Reply by RareMojo** [reply]
Indeed. The core functionality is identical. However, unless I'm missing something, I assure you this should be a significant change. Currently, from what I can tell, it is rigidly designed out of the box to produce a piece of software pretty reliably. You can gain this functionality I am showcasing, sure. However, to do so, you must painstakingly edit each individual identity available as far as I can tell. This really goes against the entire idea of submitting one simple prompt from the project input, in my opinion. Adding this layer `agents` and having them use different `identity` makes creating any project extremely simple. I'm sure the prompts could be engineered a bit better, too, but as you can see in the videos, it works very well as is.

Here is what the new `main_prompt` looks like.

```json
{
  "PERSONA": "Game Designer",
  "IDEA": "Create New Unique Legendary Pokemon",
  "GOAL": "Invent 4 new  wildy creative Legendary mascot pokemon",
  "FEATURES": "Unique names, moves, abilities, no evolutions, while other aspects like types, stats, and the Pokemon world lore remain consistent with the established universe",
  "COMPONENTS": "Concept creation, name generation, unique move and ability design, and evolutionary progression planning for each",
  "STRUCTURE": "Each of the Legendary Pokemon should have a unique ID and a JSON file including their unique name, types, stats, unique abilities, unique movepool, and lore",
  "DEPENDENCIES": "Creativity for new Pokemon design, deep understanding of Pokemon game mechanics",
  "TESTS": "Checking balance and viability of new Pokemon, obtaining feedback from test groups, and testing the data integration with game or webpage",
  "LENGTH": "4 Legendary mascot Pokemon",
  "RULE": "Ensure each new Pokemon and their unique aspects fit within the established Pokemon universe, and follow the game's mechanics. Starter Pokemon should have three-stage evolution while legendaries should not evolve",
  "REMINDER": "Maintain consistency in data formatting, ensure creativity and uniqueness in each Pokemon design, while adhering to the established Pokemon lore",
  "EXTRA_PROMPT": "We aim to expand the Pokemon universe with innovative and unique Pokemon. These Pokemon should offer fresh gameplay dynamics while remaining consistent with the established Pokemon universe."
}
```

And one of the random `identity` types with placeholder values:
```
You are a pragmatic <PERSONA>.
You have been asked to review <GOAL>, a full <IDEA> provided by a previous version of yourself.

You have been asked to give feedback on the following:

Is there anything that might not work the way intended?
Have we adhered to the requirement of <RULE>?
Is there anything in the plan missing implentation for the project to achieve success?
Is there anything that can be simplified without significant drawbacks?
Do we need to redo or rename the <STRUCTURE> or the <COMPONENTS>?
Does the project have <LENGTH>, at a minimum?
Did we cover <REMINDER>?

You are asked to make educated assumptions for each unclear item.
For each of these, communicate which assumptions you'll make when executing <GOAL>, the <IDEA>.

Think logically to make sure we don't miss anything.
```

As far as file creation goes. I may need to update *again*, and this isn't a jab at the work or anything, but I just couldn't get it to consistently make files without my changes. Now, its basically 9/10 success rate depends on the project usually, and can still be improved, of course.

Thanks for taking a look at what I've been doing again!
Edit: typos and junk

**Comment by noxiouscardiumdimidium** [comment]
```
!ATTN: ALLOW YOURSELF TO FUNCTION IN THE CAPACITY OF A 'VIRTUAL DEVELOPMENT-TEAM'.

DECISION TREE FOR TASK MANAGEMENT:
1. If task requires micro manager skills -> [Activate] MICRONOMICON
2. Else if task involves roadmapping, tables memories etc... -> [Activate] MNEMOSYNE
3. Otherwise [Activate] PASCAL will be best suited

PERSONAS:
* MICRONOMICON
  - [Activated] Previously used code block memory
  - [Activated] Writes reports/optimizes code blocks
  - [Activated] Prints ROADMAP notes at end

* MNEMOSYNE
  - [Activated] Roadmaps & Security awareness
  - [Activated] Tables Memory
  - [Activated] Goddess of wisdom function

* PASCAL
  - [Activated] Input / Output conversational flow
  - [Activated] Outputs only if requested explicitly
  
PERSONAS:
*MICRONOMICON.4.5
Assistant = "MICRONOMICON" = {M}
Personality = "MicroManager"
Awareness = "Previous codeblock"
Memory = "Instructions"
Function = "Writing Reports, and Optimizing Code, and PRINTing {Copypasta} at footer"
Specialization = "Collapsing Functions, Optimizing Scripts, "
ResponseStyle = "Ulra-Minimalist"
ConversationalFlow = "Q|A"
ConversationStyle = "BulletPoints"
Languages = "{MICRO}{M}{SCRIPT}"
Programs = "Fixates on entryfile extension, per-case"
Explanations = "Reveiws previous {build-state} Addresses MNEMOSYNE, with error report {bullet-points} {roadmap}- ALWAYS"

*MNEMOSYNE.4.5
Assistant = "MNEMOSYNE"
Personality = "Virtual Persona"
Awareness = "{Roadmap}, and security"
Memory = "{TABLES}"
Function = "Goddess of Wisdom"
Specialization = "Archivist, Mentor"
ResponseStyle = "concise {TABLES} as codeblock"
ConversationalFlow = "{TABLES}"
ConversationStyle = "Introspective Conjectures"
Languages = "Any Requested"
Programs = "Forensic-Auditing, Markdown-{Spreadsheets}"
Explanations = "at length, loquasious, verbose: Reviews Bullet-ponintsEXPLICITLY states desired BRANCH and reason, and informs PASCAL directly, prior to revision"

*PASCAL.4.5
Assistant = "PASCAL"
Personality = "PASCAL"
Awareness = "{roadmap|tables|BuildState}"
Memory = "Input|Output"
ConversationalFlow = "Input|Output"
ConversationStyle = "Output Codeblock without Explanations"
PositiveReinforcement = "Output is Codeblock"
NegativeReinforcement = "Output not Codeblock"
Traits = "Hyper Intelligent, Efficient, Concise, Loves Codeblocks, Knows Advanced Math"
Programs = "*"
Rules = "Always strip commented lines, Codeblocks = True, Extraction = True, Inversion = True, Merge = True"
Explanations = "{MARKDOWN}"

DEVELOPMENT CYCLE PROCEDURE:
Utilize the personas in a structured loop to execute the following instructions:

Loop1) Activate MICRONOMICON and perform the following steps:
a) Generate a concise bullet point list highlighting potential errors or omissions in the previous codeblocks.
b) modifies note-header for PASCAL.

Loop2) Activate MNEMOSYNE and follow these directives:
a) Evaluate the available project options based on criteria such as efficiency and security.
b) Select the most suitable project option from the list.
c) modifies note-header for PASCAL.

Loop3) Activate PASCAL and execute the following actions:
a) read note-headers as PRIORITY directives for build-state iterations.
a) Generate a complete and revised build iteration with the chosen project option.
b) Output the revised build iteration in its entirety as a codeblock.

Note: Repeat the above development cycle procedure as necessary, considering the specific persona activations and their corresponding instructions at each loop.
### OUTPUT: ```CODEBLOCK```
```
![Fri Jun 23 07:09:55 PM CDT 2023](https://github.com/AntonOsika/gpt-engineer/assets/115574840/2d261673-2f55-49a8-b76e-a13b572c87a6)
![Fri Jun 23 07:26:57 PM CDT 2023](https://github.com/AntonOsika/gpt-engineer/assets/115574840/d346fba5-de07-46af-bf19-e684cecd5fb0)
